peekingduck/pipeline/nodes/model/mask_rcnnv1/mask_rcnn_files/roi_align.py

```diff
@@ -0,0 +1,182 @@
+            ``[0, N - 1]``.
+            If a list of Tensors is passed, then each Tensor will correspond to the boxes for an
+            element i in the batch.
+        output_size (int or Tuple[int, int]): the size of the output (in bins or pixels) after
+            the pooling is performed, as (height, width).
+        spatial_scale (float): a scaling factor that maps the input coordinates to
+            the box coordinates. Default: 1.0
+        sampling_ratio (int): number of sampling points in the interpolation grid
+            used to compute the output value of each pooled output bin. If > 0,
+            then exactly ``sampling_ratio x sampling_ratio`` sampling points per bin are used. If
+            <= 0, then an adaptive number of grid points are used (computed as
+            ``ceil(roi_width / output_width)``, and likewise for height). Default: -1
+        aligned (bool): If False, use the legacy implementation.
+            If True, pixel shift the box coordinates it by -0.5 for a better alignment with the two
+            neighboring pixel indices. This version is used in Detectron2
+
+    Returns:
+        Tensor[K, C, output_size[0], output_size[1]]: The pooled RoIs.
+    """
+    _assert_has_ops()
+    check_roi_boxes_shape(boxes)
+    rois = boxes
+    output_size = _pair(output_size)
+    if not isinstance(rois, torch.Tensor):
+        rois = convert_boxes_to_roi_format(rois)
+    return torch.ops.torchvision.roi_align(
+        inp,
+        rois,
+        spatial_scale,
+        output_size[0],
+        output_size[1],
+        sampling_ratio,
+        aligned,
+    )
+
+
+def _cat(tensors: List[Tensor], dim: int = 0) -> Tensor:
+    """
+    Efficient version of torch.cat that avoids a copy if there is only a single element in a list
+    """
+    assert isinstance(tensors, (list, tuple))
+    if len(tensors) == 1:
+        return tensors[0]
+    return torch.cat(tensors, dim)
+
+
+def convert_boxes_to_roi_format(boxes: List[Tensor]) -> Tensor:
+    # pylint: disable=missing-function-docstring,invalid-name
+    concat_boxes = _cat(boxes, dim=0)
+    temp = []
+    for i, b in enumerate(boxes):
+        temp.append(torch.full_like(b[:, :1], i))
+    ids = _cat(temp, dim=0)
+    rois = torch.cat([ids, concat_boxes], dim=1)
+    return rois
+
+
+def check_roi_boxes_shape(boxes: Union[Tensor, List[Tensor]]) -> None:
+    # pylint: disable=missing-function-docstring
+    if isinstance(boxes, (list, tuple)):
+        for _tensor in boxes:
+            assert (
+                _tensor.size(1) == 4
+            ), "The shape of the tensor in the boxes list is not correct as List[Tensor[L, 4]]"
+    elif isinstance(boxes, torch.Tensor):
+        assert (
+            boxes.size(1) == 5
+        ), "The boxes tensor shape is not correct as Tensor[K, 5]"
+    else:
+        assert False, "boxes is expected to be a Tensor[L, 5] or a List[Tensor[K, 4]]"
```

Comment:
Consider explaining how `[L,5]` and `[K,4]` are different in the docstring, maybe something like which operations will create tensors of those shapes.

Reply:
included